src/Castle.Core.Tests/MethodComparerTestCase.cs

```diff
@@ -355,5 +355,28 @@ public void Compare_inherited_double_nested_return_method()
 			Assert.IsTrue(mc.Equals(typeof(Base).GetMethod("GenericMethod7"),
 				typeof(Inherited).GetMethod("GenericMethod7")));
 		}
+
+		[Test]
+		public void Compare_two_method_overloads_with_generic_arg_types()
+		{
+			var mc = MethodSignatureComparer.Instance;
+			var methods = typeof(IHaveOverloadedGenericMethod).GetMethods();
+			var firstOverload = methods[0];
+			var secondOverload = methods[1];
+
+			// The overloads must obviously have different signatures, or we could never even have
+			// compiled this test code successfully. The arguments must have a different type:
+			Assert.IsFalse(mc.Equals(firstOverload, secondOverload));
+		}
+
+		private interface IHaveOverloadedGenericMethod
+		{
+			void GenericMethod<T>(GenericClass1<T> arg);
+			void GenericMethod<T>(GenericClass2<T> arg);
+		}
+
+		private class GenericClass1<T> { }
+
+		private class GenericClass2<T> { }
```

Comment:
I felt it was important to keep the two tests separate:

* The unit test here deals with some internal cogwheel;
* The regression unit test acts at the level of the public API.

Yet, declaring the same test types twice seems somewhat repetitive. Do you mind? (If so, where would be the best place to put them?)

Reply:
> Yet, declaring the same set of test types (the interface, plus GenericClass1<T> and GenericClass2<T>) twice seems somewhat repetitive. Do you mind? (If so, where would be the best place to put them?)

Two tests sounds fine to me, means if the internal implementation changes we'll at least still have the regression one.